man/man1/tau_merge.1
.\" ** You probably do not want to edit this file directly **
.\" It was generated using the DocBook XSL Stylesheets (version 1.69.1).
.\" Instead of manually editing it, you probably should edit the DocBook XML
.\" source for it and then use the DocBook XSL Stylesheets to regenerate it.
.TH "TAU_MERGE" "1" "12/12/2008" "" "Tools"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
tau_merge \- combine multiple node and or thread TAU tracefiles into a merged tracefile
.SH "SYNOPSIS"
.HP 10
\fBtau_merge\fR [\fB\-a\fR] [\fB\-r\fR] [\fB\-n\fR] [\fB\-e\fR\ \fIeventfile_list\fR] [\fB\-m\fR\ \fIoutput_eventfile\fR] {\fItracefile_list\fR} [{\fIoutput_tracefile\fR} {\fI\ \-\ \fR}]
.SH "DESCRIPTION"
.PP
tau_merge is generated when TAU is configured with the \-TRACE option.
.PP
This tool assembles a set of tau trace and event files from multiple multiple nodes or threads across a program's execution into a single unified trace file. Many TAU trace file tools operate on merged trace files.
.PP
Minimally, tau_merge must be invoked with a list of unmerged trace files followed by the desired name of the merged trace file or the \- flag to send the output to the standard out. Typically the list can be designated by giving the shared name of the trace files to be merged followed by desired range of thread or node designators in brackets or the wild card character '*' to encompass variable thread and node designations in the filename (trace.A.B.C.trc where A, B and C are the node, context and thread numbers respectively). For example tautrace.*.trc would represent all tracefiles in a given directory while tautrace.[0\-5].0.0.trc would represent the tracefiles of nodes 0 through 5 with context 0 and thread 0.
.PP
tau_merge will generate the specified merged trace file and an event definition file, tau.edf by default.
.PP
The event definition file can be given an alternative name by using the '\-m' flag followed by the desired filename. A list of event definition files to be merged can be designated explicitly by using the '\-e' flag followed by a list of unmerged .edf files, specified in the same manner as the trace file list.
.PP
If computational resources are insufficient to merge all trace and event files simultaneously the process may be undertaken hierarchically. Corresponding subsets of the tracefiles and eventfiles may be merged in sequence to produce a smaller set of files that can then be to merged into a singular fully merged tracefile and eventfile. E.g. for a 100 node trace, trace sets 1\-10, 11\-20, ..., 91\-100 could be merged into traces 1a, 2a, ..., 10a. Then 1a\-10a could be merged to create a fully merged tracefile.
.SH "OPTIONS"
.PP
\fB\-e\fR
eventfile_list explicitly define the eventfiles to be merged
.PP
\fB\-m\fR
output_eventfile explicitly name the merged eventfile to be created
.PP
\fB\-\fR
send the merged tracefile to the standard out
.PP
\fB\-a\fR
adjust earliest timestamp time to zero
.PP
\fB\-r\fR
do not reassemble long events
.PP
\fB\-n\fR
do not block waiting for new events. By default tau_merge will block and wait for new events to be appended if a tracefile is incomplete. This command allows offline merging of (potentially) incomplete tracefiles.
.SH "EXAMPLES"
.PP
To merge all TAU tracefiles into app.trc and produce a merged tau.edf eventfile:
.sp
.nf
tau_merge *.trc app.trc
      
.fi
.sp
To merge all eventfiles 0\-255 into ev0_255merged.edf and TAU tracefiles for nodes 0\-255 into the standard out:
.sp
.nf
tau_merge \-e events.[0\-255].edf \-m ev0_255merged.edf \\
  tautrace.[0\-255].*.trc \-
      
.fi
.sp
To merge eventfiles 0, 5 and seven info ev057.edf and tau tracefiles for nodes 0, 5 and 7 with context and thread 0 into app.trc:
.sp
.nf
tau_merge \-e events.0.edf events.5.edf events.7.edf \-m ev057.edf \\
  tautrace.0.0.0.trc tautrace.5.0.0.trc tautrace.7.0.0.trc app.trc
      
.fi
.sp
.SH "SEE ALSO"
.PP
tau_convert
.PP
tau2profile
.PP
tau2vtf
.PP
tau2elg
.PP
tau2slog2


Changes
*****************************************************************************
**                      TAU Portable Profiling Package                     **
**                      http://tau.uoregon.edu                             **
*****************************************************************************
**    Copyright 1997-2008						   **
**    Department of Computer and Information Science, University of Oregon **
**    Advanced Computing Laboratory, Los Alamos National Laboratory        **
*****************************************************************************

Change log:
------------
Version 2.18 changes (from 2.17):
1. Added support for PIN based runtime instrumentation for Windows.
2. Added support for compiler based instrumentation for PGI and IBM compilers.
3. Added support for thread-safe throttling of events.
4. Added support for reading annotated snapshot perfsuite profiles generated by TAU.
5. Enhanced support for GNU based compiler instrumentation to support instrumented shared objects. 
6. Added support for parsing C99 code using PDT. 
7. Python API enhancements include support for setting node, data purging and exit.
8. tau_ompcheck now supports more OpenMP directives. 
9. PerfExplorer now includes a CQoS classifier and a gap computing module.
10. Compiler based instrumentation supports selective instrumentation for the file level. 
11. Added a -bfd=download configure option which will download and build binutils with -fPIC for compiler based instrumentation. 
12. Added support for Intel v11 compilers. 
13. Snapshot enhancements (read snapshot.*.*.* files, paraprof -f snapshot)
14. Better configuration support for SiCortex, IBM BG/P and Cray XT5 (PGI, Pathscale, GNU), Intel compilers for Apple.
15. Added support for preloading pthread calls enable with -useropt=-DTAU_PTHREAD_PRELOAD
16. Added support for -DISABLESHARED configure option that does not build libTAU.so.

Version 2.17 changes (from 2.16):
1. Added support for IBM BG/P (-arch=bgp).
2. Added a new tool for generating wrapper libraries, tau_wrap.
3. Improvements in Eclipse plugin for external tool support.
4. Improvements in paraprof and perfexplorer.
5. Improvements for SiCortex support and tauex. 
6. Added support for atomic events in TAU library layered over VampirTrace.
7. Added a Posix I/O wrapper (-iowrapper) for tracking volume and bandwidth of I/O.
8. Added an MPI wrapper library for Windows Cluster 2003. 
9. Added support for Scalasca 1.0. Works with both Kojak and Scalasca. 
10. Added Opari in TAU.
11. Added IBM BG/P metadata for torus node information in profiles.
12. PerfExplorer adds support for user-defined events and improvements in custom charts.
13. Posix I/O tracking implemented without need for enabling profiling (for tracing).
14. Improved tau_inc.pl for generating include lists for Scalasca/Kojak based on callpath profiling. 
15. Eclipse TAU plugin has support for two stage communication analysis.
16. Added -BGPTIMERS for IBM BG/P. Compatible with -BGLTIMERS.
17. Env vars TAU_VERBOSE, TAU_SYNCHRONIZE_CLOCKS, TAU_PROFILE_FORMAT (snapshot)
18. GCC 4.3.0 compatibility
19. Added bandwidth and bytes written info for MPI I/O write routines.
20. Added support for GNU, PathScale and PGI compilers on Cray XT systems [ORNL].
21. ParaProf can now generate selective instrumentation files.
22. TAU_THROTTLE = 0 disables throttling of events. Use TAU_VERBOSE=1 to see it.
23. perfdmf_configure now stores weka.jar files in ~/.ParaProf directory.
24. Added support for DyninstAPI 5.2.
25. Bug fixes for tau_instrumentor, context events, and tau2slog2.
26. Added support for pointer based profiling API (examples/profilercreate/README) [LSU].
27. -spec option for tau_instrumentor allows generic timer instrumentation support [FZJ].
28. Paraprof allows new windows for multiple metrics to compare data [SiCortex].
29. Posix I/O tracking now uses context events instead of user-defined events.
30. Added support for compiler based instrumentation for Intel 9.1, 10.x, GNU, and PathScale compilers. 
31. Added extensions in PerfExplorer to support CQoS analysis, drawing charts from script [CCA]. 
32. Bug fixes in paraprof for selective instrumentation, printer support [NASA].
33. taucxx, taucc, tauf90 now use -optCompInst by default, tau_[cxx,cc,f90].sh use -optPDTInst by default.
34. Added -opari support in installtau. 
35. Fixes for IBM BGL/BGP configuration.
36. Added support for tracking memory utilization and headroom in Python [ALCF].



Version 2.16 changes (from 2.15):
1. Added a new tool for correcting network time drifts in traces (tau_timecorrect).
2. Added support for an Eclipse analysis wizard and a graphical instrumentor.
3. Added support for 3D stereo visualizations in ParaProf.
4. Added support for a source browser in ParaProf.
5. Added support for generating source code information in tau_instrumentor.
6. Added support for Perflib based instrumentation and perf2tau [Jeff Brown, LANL]. 
7. Updated KTAU support in TAU for registering fork for kernel profiling [ANL]. 
8. Added tau_validate tool for checking if the TAU library is built correctly [UTK].
9. Added support for loading multiple ppk files in paraprof on the commandline [ORNL]. 
10. Enhancements in ParaProf and PerfExplorer for using metadata. [PERI]
11. Added support for capturing date and other cpu information in profiles. [LLNL]
12. Added support for Vampirtrace [LLNL]. 
13. Added support for Scalasca 0.5, and KOJAK 2.2 [FZJ, UTK]. 
14. Enhancements to Eclipse PTP plugin to support PAPI counter selection. [UTK]
15. Supports PDT v3.10 with EDG v3.8 C++/C parsers [LLNL].
16. Added support for application signatures [RENCI].
17. Added support for SiCortex Linux platform [SiCortex].

utils/tau_merge.cpp

#include "TAU_tf.h"

#include <iostream>
#include <vector>
#include <map>
#include <string>

using namespace std;

#include <float.h>


Ttf_FileHandleT theFile;
map <string,int> eventMap; 
map <string,int> groupMap; 
map <string,int> userEventMap;

int DefThread(void *userData, unsigned int nodeToken, unsigned int threadToken, const char *threadName);


int DefStateGroup(void *userData, unsigned int stateGroupToken, const char *stateGroupName);
int DefState(void *userData, unsigned int stateToken, const char *stateName, unsigned int stateGroupToken);
int DefUserEvent(void *userData, unsigned int userEventToken,
		 const char *userEventName, int monotonicallyIncreasing);


int EventTrigger(void *userData, double time,
		 unsigned int nodeToken, unsigned int threadToken, unsigned int userEventToken,
		 //double userEventValue)
		 long long userEventValue);


int SendMessage(void *userData, double time,
		unsigned int sourceNodeToken,
		unsigned int sourceThreadToken,
		unsigned int destinationNodeToken,
		unsigned int destinationThreadToken, unsigned int messageSize, unsigned int messageTag, unsigned int messageComm);

int RecvMessage(void *userData, double time,
		unsigned int sourceNodeToken,
		unsigned int sourceThreadToken,
		unsigned int destinationNodeToken,
		unsigned int destinationThreadToken, unsigned int messageSize, unsigned int messageTag, unsigned int messageComm);

int EnterState(void *userData, double time, unsigned int nodeid, unsigned int tid, unsigned int stateid);
int LeaveState(void *userData, double time, unsigned int nodeid, unsigned int tid, unsigned int stateid);
int EndTrace(void *userData, unsigned int nodeToken, unsigned int threadToken);



// an interface for the various record types
class Record {
	public:
		Record(double time) {
			this->time = time;	
		}
		double getTime() {
			return time;	
		}
		virtual void writeRecord(Ttf_FileHandleT tFile) = 0;
		
	protected:
		double time;
};


class EnterStateRecord : public Record {
	public:
		EnterStateRecord(double time, int nid, int tid, int stateToken) : 
					Record(time) {
			this->stateToken = stateToken;
			this->nid = nid;
			this->tid = tid;
		}

		void writeRecord(Ttf_FileHandleT tFile) {
			printf ("writing EnterState record: state=%d, nid=%d, tid=%d\n",stateToken,nid,tid);
			Ttf_EnterState(tFile, time, nid, tid, stateToken);
		}

	private:
		int nid;
		int tid;
		int stateToken;
};

class LeaveStateRecord : public Record {
	public:
		LeaveStateRecord(double time, int nid, int tid, int stateToken) : 
					Record(time) {
			this->stateToken = stateToken;
			this->nid = nid;
			this->tid = tid;
		}

		void writeRecord(Ttf_FileHandleT tFile) {
			Ttf_LeaveState(tFile, time, nid, tid, stateToken);
		}

	private:
		int nid;
		int tid;
		int stateToken;
};


class TriggerRecord : public Record {
	public:
		TriggerRecord(double time, int nid, int tid, int userEventToken, x_uint64 value) : Record(time) {
			this->nid = nid;
			this->tid = tid;
			this->userEventToken = userEventToken;
			this->value = value;
		}
		
		void writeRecord(Ttf_FileHandleT tFile) {
			Ttf_EventTrigger(tFile, time, nid, tid, userEventToken, value )	;
		}
	
	private:
		int nid;
		int tid;
		int userEventToken;
		double value;
};



class MessageRecord : public Record {
	public:
		MessageRecord(double time,
		unsigned int sourceNodeToken,
		unsigned int sourceThreadToken,
		unsigned int destinationNodeToken,
		unsigned int destinationThreadToken, unsigned int messageSize, unsigned int messageTag, int messageComm, bool send) : Record(time) {
			this->sourceNid = sourceNodeToken;
			this->sourceTid = sourceThreadToken;
			this->destNid = destinationNodeToken;
			this->destTid = destinationThreadToken;
			this->size = messageSize;
			this->tag = messageTag;
			this->comm = messageComm;
			this->send = send;
		}

		void writeRecord(Ttf_FileHandleT tFile) {
		if (send) {
				Ttf_SendMessage(tFile, time, sourceNid, sourceTid, destNid, destTid, size, tag, comm);
			} else {
				Ttf_RecvMessage(tFile, time, sourceNid, sourceTid, destNid, destTid, size, tag, comm);
			}
		}
		
	private:
		unsigned int sourceNid, sourceTid, destNid, destTid, size, tag, comm;
		bool send;
	
	
};


class Trace {


  public:
    Trace(Ttf_FileHandleT tFile) : esr(0,0,0,0), lsr(0,0,0,0), tr(0,0,0,0,0), mr(0,0,0,0,0,0,0,0,0)
{
    this->tFile = tFile;
	this->finished = false;
	this->lastRecord = NULL;
    callbacks.DefClkPeriod = NULL;
    callbacks.DefThread = DefThread;
    callbacks.DefStateGroup = DefStateGroup;
    callbacks.DefState = DefState;
    callbacks.DefUserEvent = DefUserEvent;
    callbacks.EventTrigger = EventTrigger;
    callbacks.EndTrace = EndTrace;
    callbacks.EnterState = EnterState;
    callbacks.LeaveState = LeaveState;
    callbacks.SendMessage = SendMessage;
    callbacks.RecvMessage = RecvMessage;

} 

Record* getLastRecord() {
	return lastRecord;	
}

bool isFinished() {
	return finished;	
}

int readNumEvents(int num) {
	printf ("%p: reading data\n", this);
	callbacks.UserData = this; // must do this here, otherwise we might have the wrong pointer
    int recs_read = Ttf_ReadNumEvents(tFile, callbacks, num);
    return recs_read;
}

int defThread(unsigned int nodeToken, unsigned int threadToken, const char *threadName) {
    printf("defThread!\n");
    nodeMap[pair<int,int> (nodeToken,threadToken)] = false;
    Ttf_DefThread(theFile, nodeToken, threadToken, threadName);
    return 0;
}

int defStateGroup(unsigned int stateGroupToken, const char *stateGroupName) {
    printf("defStateGroup : %d : %s\n", stateGroupToken, stateGroupName);

	map<string, int>::iterator it =	groupMap.find(stateGroupName);
	if (it == groupMap.end()) { // not found, add it
		int globalId = groupMap.size();
		groupMap[stateGroupName] = globalId;
		localGroupMap[stateGroupToken] = globalId;
		Ttf_DefStateGroup(theFile, stateGroupName, globalId);
	} else {
		localGroupMap[stateGroupToken] = (*it).second;
	}

    return 0;
}

int defState(unsigned int stateToken, const char *stateName, int stateGroupToken) {
    printf("defState : %d : %s\n", stateToken, stateName);

	map<string, int>::iterator it =	eventMap.find(stateName);
	if (it == eventMap.end()) { // not found, add it
		printf ("didn't find %s in the map\n", stateName);
		int globalId = eventMap.size()+1;
		eventMap[stateName] = globalId;
		localEventMap[stateToken] = globalId;
		
		string crappedName = stateName;
		crappedName = crappedName.substr(1,crappedName.size()-2);
		Ttf_DefState(theFile, globalId, crappedName.c_str(), localGroupMap[stateGroupToken]);
	} else {
		printf ("found %s in the map\n", stateName);
		localEventMap[stateToken] = (*it).second;
	}

    
    return 0;
}


int defUserEvent(unsigned int userEventToken, const char *userEventName, int monotonicallyIncreasing) {
    printf("defUserEvent : %d : %s\n", userEventToken, userEventName);

	map<string, int>::iterator it =	userEventMap.find(userEventName);
	if (it == userEventMap.end()) { // not found, add it
		int globalId = userEventMap.size()+1;
		userEventMap[userEventName] = globalId;
		localUserEventMap[userEventToken] = globalId;
		
		string crappedName = userEventName;
		crappedName = crappedName.substr(1,crappedName.size()-2);
		Ttf_DefUserEvent(theFile, globalId, crappedName.c_str(), monotonicallyIncreasing);
	} else {
		localUserEventMap[userEventToken] = (*it).second;
	}

    return 0;
}

int eventTrigger(double time, unsigned int nodeToken, unsigned int threadToken, unsigned int userEventToken,
		 //double userEventValue)
		 long long userEventValue) {
    printf("eventTrigger!\n");
    tr = TriggerRecord(time, nodeToken, threadToken, userEventToken, userEventValue);
    lastRecord = &tr;
    return 0;
}
int sendMessage(double time,
		unsigned int sourceNodeToken,
		unsigned int sourceThreadToken,
		unsigned int destinationNodeToken,
		unsigned int destinationThreadToken, unsigned int messageSize, unsigned int messageTag, 
		unsigned int messageComm) {
    printf("sendMessage!\n");
    mr = MessageRecord(time, sourceNodeToken, sourceThreadToken, destinationNodeToken, 
				       destinationThreadToken, messageSize, messageTag, messageComm, true);
    lastRecord = &mr;
    return 0;
}

int recvMessage(double time,
		unsigned int sourceNodeToken,
		unsigned int sourceThreadToken,
		unsigned int destinationNodeToken,
		unsigned int destinationThreadToken, unsigned int messageSize, unsigned int messageTag, 
		unsigned int messageComm) {
    printf("recvMessage!\n");
    mr = MessageRecord(time, sourceNodeToken, sourceThreadToken, destinationNodeToken, 
				       destinationThreadToken, messageSize, messageTag, messageComm, false);
    lastRecord = &mr;
    return 0;
}
int enterState(double time, unsigned int nodeid, unsigned int tid, unsigned int stateid) {
	printf("EnterState callback: state=%d, nid=%d, tid=%d\n", stateid,nodeid,tid);
    esr = EnterStateRecord(time,nodeid,tid,localEventMap[stateid]);
    lastRecord = &esr;
    return 0;
}
int leaveState(double time, unsigned int nodeid, unsigned int tid, unsigned int stateid) {
    printf("leaveState!\n");
    lsr = LeaveStateRecord(time,nodeid,tid,localEventMap[stateid]);
	lastRecord = &lsr;
    return 0;
}

int endTrace(unsigned int nodeToken, unsigned int threadToken) {
    printf("%p: endTrace!\n",this);
    nodeMap[pair<int,int> (nodeToken,threadToken)] = true;

	map< pair<int, int> , bool, less< pair<int,int> > >::iterator it;
	printf ("map size = %d\n", nodeMap.size());
	finished = true;
	for (it = nodeMap.begin(); it != nodeMap.end(); it++) {
		if ((*it).second == false) {
			finished = false;	
		}
	}
	return 0;
}


int writeRecord(Ttf_FileHandleT tFile) {
	printf ("%p: writing data\n", this);
	lastRecord->writeRecord(tFile);
	lastRecord = NULL;	
	return 0;
}

private:

	Ttf_FileHandleT tFile;
	Ttf_CallbacksT callbacks;
	Record *lastRecord;
	bool finished;

	EnterStateRecord esr;
	LeaveStateRecord lsr;
	TriggerRecord tr;
	MessageRecord mr;

	map< pair<int, int> , bool, less< pair<int,int> > > nodeMap;

	map< int, int > localEventMap; // map local event id's to global
	map< int, int > localGroupMap; // map local group id's to global
	map< int, int > localUserEventMap; // map local event id's to global

};


int DefThread(void *userData, unsigned int nodeToken, unsigned int threadToken, const char *threadName)
{
    return ((Trace *) userData)->defThread(nodeToken, threadToken, threadName);
}
int DefStateGroup(void *userData, unsigned int stateGroupToken, const char *stateGroupName)
{
    return ((Trace *) userData)->defStateGroup(stateGroupToken, stateGroupName);
}

int DefState(void *userData, unsigned int stateToken, const char *stateName, unsigned int stateGroupToken)
{
    return ((Trace *) userData)->defState(stateToken, stateName, stateGroupToken);
}


int DefUserEvent(void *userData, unsigned int userEventToken,
		 const char *userEventName, int monotonicallyIncreasing)
{
    return ((Trace *) userData)->defUserEvent(userEventToken, userEventName, monotonicallyIncreasing);
}
int EventTrigger(void *userData, double time,
		 unsigned int nodeToken, unsigned int threadToken, unsigned int userEventToken,
		 //double userEventValue)
		 long long userEventValue)
{
    return ((Trace *) userData)->eventTrigger(time, nodeToken, threadToken, userEventToken, userEventValue);

}
int SendMessage(void *userData, double time,
		unsigned int sourceNodeToken,
		unsigned int sourceThreadToken,
		unsigned int destinationNodeToken,
		unsigned int destinationThreadToken, unsigned int messageSize, unsigned int messageTag, unsigned int messageComm)
{
    return ((Trace *) userData)->sendMessage(time, sourceNodeToken, sourceThreadToken, destinationNodeToken,
					     destinationThreadToken, messageSize, messageTag, messageComm);
}
int RecvMessage(void *userData, double time,
		unsigned int sourceNodeToken,
		unsigned int sourceThreadToken,
		unsigned int destinationNodeToken,
		unsigned int destinationThreadToken, unsigned int messageSize, unsigned int messageTag, unsigned int messageComm)
{
    return ((Trace *) userData)->recvMessage(time, sourceNodeToken, sourceThreadToken, destinationNodeToken,
					     destinationThreadToken, messageSize, messageTag, messageComm);
}

int EnterState(void *userData, double time, unsigned int nodeid, unsigned int tid, unsigned int stateid)
{
    return ((Trace *) userData)->enterState(time, nodeid, tid, stateid);
}
int LeaveState(void *userData, double time, unsigned int nodeid, unsigned int tid, unsigned int stateid)
{
    return ((Trace *) userData)->leaveState(time, nodeid, tid, stateid);
}
int EndTrace(void *userData, unsigned int nodeToken, unsigned int threadToken)
{
    return ((Trace *) userData)->endTrace(nodeToken, threadToken);
}


bool dontBlock = false;
bool adjustToZero = false;


void usage(char *name)
{
    fprintf(stderr,
	    "Usage: %s [-a] [-n] [-e eventedf*] [-m mergededf] inputtraces* (outputtrace|-) \n\n", name);
    fprintf(stderr, "Options:\n\n");
    fprintf(stderr, "-a : adjust first time to zero\n");
    fprintf(stderr, "-n : do not block waiting for new events. Offline merge\n");
    fprintf(stderr, "-e <files> : provide a list of event definition files corresponding to traces\n");
    fprintf(stderr, "-m <mergededf> : specify the name of the merged event definition file\n\n");
    fprintf(stderr, "Note: %s assumes edf files are named events.<nodeid>.edf and \n", name);
    fprintf(stderr, "      generates a merged edf file tau.edf unless -m is specified\n\n");
    fprintf(stderr, "e.g., > %s tautrace.*.trc app.trc\n", name);
    fprintf(stderr, "e.g., > %s -e events.[0-255].edf -m ev0_255merged.edf tautrace.[0-255].*.trc app.trc\n\n",
	    name);
    exit(1);
}


int getNodeId(char *name)
{
    // name is something like tautrace.32.0.4.trc
    // we want to return 32

    char buffer[1024];
    unsigned int i = 0;
    while (i < strlen(name) && name[i] != '.') {
	i++;
    }
    i++;
    int bufpos = 0;
    while (i < strlen(name) && name[i] != '.') {
	buffer[bufpos++] = name[i];
	i++;
    }
    buffer[bufpos] = 0;

    return atoi(buffer);
}

int cannot_get_enough_fd(int need)
{
#ifdef TAU_WINDOWS
    return false;		// no getdtablesize() in windows
#else
# if defined(__hpux) || defined(sun)
    /* -- system supports get/setrlimit (RLIMIT_NOFILE) -- */
    struct rlimit rlp;

    getrlimit(RLIMIT_NOFILE, &rlp);
    if (rlp.rlim_max < need) {
	return (TRUE);
    } else if (rlp.rlim_cur < need) {
	rlp.rlim_cur = need;
	setrlimit(RLIMIT_NOFILE, &rlp);
    }
    return (FALSE);
# else
#   if defined(_SEQUENT_) || defined(sequent)
    /* -- system provides get/setdtablesize -- */
    int max = getdtablesize();
    return ((max < need) && (setdtablesize(need) != need));
#   else
    /* -- system provides only getdtablesize -- */
    int max = getdtablesize();
    return (max < need);
#   endif
# endif
#endif				/* TAU_WINDOWS */
}


extern char *optarg;
extern int optind;
char **edfNames;		/* if they're specified */
int edfCount = 0;
bool edfSpecified = false;

int main(int argc, char *argv[])
{

    char *mergedEdfFile = "tau.edf";	/* initialize it */

    int numEdfProcessed;
    int i;
    while ((i = getopt(argc, argv, "ane:m:")) != EOF) {
	switch (i) {
	case 'a':		/* -- adjust first time to zero -- */
	    adjustToZero = true;
	    break;


	case 'e':		/* -- EDF files specified on the commandline -- */
	    edfSpecified = true;
	    numEdfProcessed = 0;
	    edfNames = (char **) malloc(argc * (sizeof(char *)));
	    for (i = optind - 1; i < argc; i++) {
		if (strstr(argv[i], ".edf") != 0) {
		    //open_edf_file(argv[i], numedfprocessed, TRUE);
		    numEdfProcessed++;
		    /* store the name of the edf file so that we can re-open 
		     * it later if event files need to be re-read */
		    edfNames[edfCount] = strdup(argv[i]);
		    edfCount++;	/* increment the count */
		} else {
		    break;	/* come out of the loop! */
		}
	    }
	    optind += numEdfProcessed - 1;
	    break;

	case 'm':		/* -- name of the merged edf file (instead of tau.edf) */
	    mergedEdfFile = strdup(argv[optind - 1]);
	    break;

	case 'n':		/* -- do not block for records at end of trace -- */
	    dontBlock = true;
	    break;

	default:		/* -- ERROR -- */
	    usage(argv[0]);
	    break;

	}
    }



    int active = argc - optind - 1;
    if (cannot_get_enough_fd(active + 4)) {
	fprintf(stderr, "%s: too many input traces:\n", argv[0]);
	fprintf(stderr, "  1. merge half of the input traces\n");
	fprintf(stderr, "  2. merge other half and output of step 1\n");
	fprintf(stderr, "  Or use, \"tau_treemerge.pl\"\n");
	exit(1);
    }


	printf ("ok...\n");
	theFile = Ttf_OpenFileForOutput("out.trc","out.edf");
	if (theFile == NULL) {
		fprintf (stderr, "Error opening	trace for output\n");
		exit(1);
	}


    vector < Trace* > traces;

    // get all the traces and edfs setup
    for (i = optind; i < argc - 1; i++) {
		char *traceFilename = argv[i];
		char *edfFilename;
		if (edfSpecified) {
		    if (traces.size() >= edfCount) {
				fprintf(stderr, "Error: Not enough EDF files were specified!\n");
				exit(1);
		    }
		    edfFilename = edfNames[traces.size()];
		} else {

	    	edfFilename = (char *) malloc(2048);
		    sprintf(edfFilename, "events.%d.edf", getNodeId(traceFilename));
		}

		printf("processing trace: %s, with edf = %s\n", traceFilename, edfFilename);

		Ttf_FileHandleT tFile = Ttf_OpenFileForInput(traceFilename, edfFilename);
		Ttf_SetSubtractFirstTimestamp(tFile, false);

		if (tFile == NULL) {
		    fprintf(stderr, "Error opening trace: %s, with edf = %s\n", traceFilename, edfFilename);
		    exit(1);
		}

		Trace *trace = new Trace(tFile);

		while (trace->getLastRecord() == NULL) {
			trace->readNumEvents(1);
		}
		traces.push_back(trace);

    }

	printf ("Here now!...\n");

    if (traces.size() < 1) {
		usage(argv[0]);
    }


	bool allFinished = false;

	while (!allFinished) {
//		printf ("processing...\n");
		double lowestTime = DBL_MAX;
		int lowestIndex = -1;
		
		for (unsigned int i=0; i < traces.size(); i++) {
			if (!traces[i]->isFinished()) {
				double time = traces[i]->getLastRecord()->getTime();
				if (time < lowestTime) {
					lowestTime = time;
					lowestIndex = i;	
				}
			}
		}
		
		if (lowestIndex == -1) {
			allFinished = true;	
		} else {
	//			printf ("lowest time with %d\n", lowestIndex);
			traces[lowestIndex]->writeRecord(theFile);
			while (traces[lowestIndex]->getLastRecord() == NULL && !traces[lowestIndex]->isFinished()) {
				traces[lowestIndex]->readNumEvents(1);
			}
		}

/*
		allFinished = true;
		for (unsigned int i=0; i < traces.size(); i++) {
			if (!traces[i].isFinished()) {
				allFinished = false;	
			}
		}		
		*/
	}
	
	Ttf_CloseOutputFile(theFile);

	printf ("done!\n");
}
